File: Falling-into-deeps/src/entity/Entity.cpp
#include "gearpch.h"
#include "Entity.h"
#include "Input.h"
#include "graphics/Graphics.h"

std::map<std::string, std::function<Component*()>> s_components;

Entity::Entity(const std::string& file) : obj(LuaManager::CreateRaw())
{
	LuaManager::Push(obj, file + ".lua");

	getGlobalNamespace(obj)
		.beginClass<TransformComponent>("transform_component")
			.addFunction("move", &TransformComponent::Move)
			.addFunction("draw", &TransformComponent::Draw)
			.addFunction("collide", &TransformComponent::Collide)
			.addFunction("setMovable", &TransformComponent::setMovable)
			.addFunction("isMovable", &TransformComponent::isMovable)
			.addFunction("isMoved", &TransformComponent::isMoved)
		.endClass()
		.beginClass<Entity>("Entity")
			.addFunction("AddComponent", &Entity::AddComponent)
			.addFunction("GetTransform", &Entity::GetTransform)
		.endClass()
		.addFunction("Draw", &Gear::Draw)
		.addFunction<bool(*)(const std::string&)>("isPressed", &Gear::isKeyPressed);

	getGlobal(obj, "init")(this);
}


Entity::~Entity()
{

}

void Entity::InitComponents()
{
	RegisterComponent<TransformComponent>("transform");
	RegisterComponent<InventoryComponent>("inventory");
}

void Entity::AddComponent(const std::string & name)
{
	components[name] = s_components[name]();
}

template<class T>
void Entity::RegisterComponent(const std::string& name)
{
	s_components[name] = []() {return new T(); };
}

File: Falling-into-deeps/res/scripts/entity/entity.lua
keys = require "res/scripts/keys"

init = function(inst)
	inst:AddComponent("transform")
	inst:GetTransform():move(128, 128)
	inst:AddComponent("inventory")
end

update = function(world, inst, delta)
	if (isPressed("a")) then
		inst:GetTransform():move(-1.5, 0)
	end
	if (isPressed("d")) then
		inst:GetTransform():move(1.5, 0)
	end
	if (isPressed("w")) then
		inst:GetTransform():move(0, -1.5)
	end
	if (isPressed("s")) then
		inst:GetTransform():move(0, 1.5)
	end
	if (inst:GetTransform():isMoved()) then
		world:collide(inst:GetTransform());
	end
end

render = function(inst)
	inst:GetTransform():draw()
end
